sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -493,13 +488,20 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
     import df.sparkSession.sessionState.analyzer.{AsTableIdentifier, CatalogObjectIdentifier}
     import org.apache.spark.sql.catalog.v2.CatalogV2Implicits._
 
-    import org.apache.spark.sql.catalog.v2.CatalogV2Implicits._
     val session = df.sparkSession
+    val provider = DataSource.lookupDataSource(source, session.sessionState.conf)
```

Comment:
the `provider` here may not be the actual table provider, as `saveAsTable` can write to an existing table. Maybe we should always use v2 session catalog?

Reply:
That works for me.

Reply:
hmm. actually that causes issues if the table doesn't exist. Maybe we should use the statements instead of the logical plans?

Reply:
+1 on using statements.

Reply:
let's do it in a followup.